src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -72,6 +77,188 @@ internal class Http2FrameWriter
+
+                var observed = Http2OutputProducer.State.None;
+
+                // We don't need to check the result because it's either
+                // - true because we have a result
+                // - false because we're flushing headers
+                reader.TryRead(out var readResult);
+
+                var buffer = readResult.Buffer;
+
+                if (producer.WriteHeaders)
+                {
+                    observed |= Http2OutputProducer.State.FlushHeaders;
+                }
+
+                if (readResult.IsCanceled)
+                {
+                    observed |= Http2OutputProducer.State.Canceled;
+                }
+
+                if (readResult.IsCompleted)
+                {
+                    observed |= Http2OutputProducer.State.Completed;
+                }
+
+                // Check the stream window
+                var (actual, remainingStream) = producer.ConsumeWindow(buffer.Length);
+
+                // Now check the connection window
+                (actual, var remainingConnection) = ConsumeWindow(actual);
+
+                // Write what we can
+                if (actual < buffer.Length)
+                {
+                    buffer = buffer.Slice(0, actual);
+                }
+
+                var (hasMoreData, reschedule) = producer.Dequeue(buffer.Length, observed);
+
+                FlushResult flushResult = default;
+
+                if (readResult.IsCanceled)
+                {
+                    // Response body is aborted, complete reader for this output producer.
+                    if (producer.WriteHeaders)
+                    {
+                        // write headers
+                        WriteResponseHeaders(stream.StreamId, stream.StatusCode, Http2HeadersFrameFlags.NONE, (HttpResponseHeaders)stream.ResponseHeaders);
+                    }
+                }
+                else if (readResult.IsCompleted && stream.ResponseTrailers is { Count: > 0 } && !hasMoreData)
+                {
+                    // Output is ending and there are trailers to write
+                    // Write any remaining content then write trailers and there's no
+                    // flow control back pressure being applied (hasMoreData)
+
+                    stream.ResponseTrailers.SetReadOnly();
+                    stream.DecrementActiveClientStreamCount();
+
+                    // It is faster to write data and trailers together. Locking once reduces lock contention.
+                    flushResult = await WriteDataAndTrailersAsync(stream, buffer, producer.WriteHeaders, stream.ResponseTrailers);
+                }
+                else if (readResult.IsCompleted && producer.StreamEnded)
+                {
+                    if (buffer.Length != 0)
+                    {
+                        _log.LogCritical("{StreamId} observed an unexpected state where the streams output ended with data still remaining in the pipe.", stream.StreamId);
```

Comment:
If we're not going to bother with `DecrementActiveClientStreamCount`, we should probably also abort the connection. I get this should never happen, but still.

Reply:
As long as I don't have to fix a boatload of tests. I preserved the behavior as best I could.

Reply:
Before this was a stream-level problem. An error was thrown and the stream output loop would exit: https://github.com/dotnet/aspnetcore/blob/e9b2adb2a21cc2b0ae5b8f7e93aa4599d6f847b0/src/Servers/Kestrel/Core/src/Internal/Http2/Http2OutputProducer.cs#L438

If we preserve that behavior then the stream/connection needs to handle the broken stream. Not decremeanting the active stream count will hang a connection if this happens often enough

Alternatively, if this is something that shouldn't ever happen then aborting the connection is a simpler Just In Case.

Reply:
I'm confused. I'm 96% sure the code is as it was before. What's different about this? The old code used to complete the pipe reader and didn't abort anything. This is new behavior...

Reply:
This new loop operates at a connection level instead of stream level.

My question is, is the stream pipe reader still completed? Is the active stream count being decremented somewhere? (TBH I'm not sure where the active stream count would be decremented in this situation in old code)

At the very least I think a comment here that explains why it's ok that after logging the critical error that `DecrementActiveClientStreamCount` isn't called and the connection's not aborted.

Reply:
> My question is, is the stream pipe reader still completed?

No.

>  Is the active stream count being decremented somewhere?

When the stream completes by other means (like the request ending).

> At the very least I think a comment here that explains why it's ok that after logging the critical error that DecrementActiveClientStreamCount isn't called and the connection's not aborted.

I'm all for fixing this but I'm not planning to introduce new behavior in this PR.

Reply:
I will make sure it's the same logic that exists today.

Reply:
The pipe is cleaned up as expected. The behavior is equivalent. We can file an issue about making it abort the connection more promptly if need be.